File: rf-controller/src/nox/netapps/tests/user_event_log_test.cc
/* Copyright 2008 (C) Nicira, Inc.
 *
 * This file is part of NOX.
 *
 * NOX is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * NOX is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with NOX.  If not, see <http://www.gnu.org/licenses/>.
 */
#include <ctime>
#include <list>
#include <sys/time.h>
#include "hash_map.hh"

#include <boost/bind.hpp>
#include <boost/function.hpp>
#include <boost/shared_ptr.hpp>

#include "component.hh"
#include "storage/storage.hh" 
#include "storage/storage_util.hh"
#include "vlog.hh"
#include "tests.hh"
#include "user_event_log/user_event_log.hh" 
#include "threads/cooperative.hh"
#include "bindings_storage/bindings_storage.hh"

using namespace std;
using namespace vigil;
using namespace vigil::container;
using namespace vigil::applications;
using namespace boost;
using namespace vigil::testing;


namespace  {

static Vlog_module lg("user-event-log-test");

class UserEventLogTestCase
    : public Test_component
{
public:

    UserEventLogTestCase(const container::Context* c,
                const json_object* xml) 
        : Test_component(c), trigger_count(0), debug(false)   {
    }

    void configure(const Configuration*) {
    }

    void install() {
          resolve(uel);
          // disable auto-clearing of UEL 
          uel->set_max_num_entries(-1); 
          resolve(store); 
          resolve(b_store); 
          main_sem = new Co_sema(); 
          test_sem = new Co_sema(); 
    }

    void run_test();

    static const int MAX_ROUNDS = 5; 
private:
    
    void set_trigger(); 
    void trigger_fired(const storage::Trigger_id & tid,const storage::Row&); 
    void trigger_inserted(const storage::Result &result,
        const storage::Trigger_id &tid);
    void remove_trigger_callback(const storage::Result &result); 
    void exit_test(bool success);
    void log_entry_cb(int64_t logid, int64_t ts, const string &app, int level, 
      const string &msg, const NameList &src_names, const NameList &dst_names);
    void do_test(); 
    void clear_cb(); 

    User_Event_Log *uel; // create message 
    storage::Async_storage *store; // confirm that messages were created
    Bindings_Storage *b_store; 
    storage::Trigger_id tid1,tid2,tid3; // used to cancel trigger
    int64_t triggered_logid; 
    int trigger_count; 
    string cur_msg;
    NameList expected_src_names, expected_dst_names; 

    bool debug; 
    Co_thread thread;
    Co_sema *main_sem, *test_sem;
};

void UserEventLogTestCase::run_test() {
  thread.start(boost::bind(&UserEventLogTestCase::set_trigger, this));
  main_sem->down();
} 

// insert trigger for 'user_event_log' table
void UserEventLogTestCase::set_trigger() {

    // first clear the table
    b_store->clear(boost::bind(&UserEventLogTestCase::clear_cb,this)); 
    test_sem->down(); 

      store->put_trigger(User_Event_Log::MAIN_TABLE_NAME,true, 
         boost::bind(&UserEventLogTestCase::trigger_fired, this, _1, _2),
         boost::bind(&UserEventLogTestCase::trigger_inserted, this, _1, _2) ); 
   
      store->put_trigger(User_Event_Log::NAME_TABLE_NAME,true, 
         boost::bind(&UserEventLogTestCase::trigger_fired, this, _1, _2),
         boost::bind(&UserEventLogTestCase::trigger_inserted, this, _1, _2) ); 
         
      // so we know when bindings operations have completed
      store->put_trigger(Bindings_Storage::NAME_TABLE_NAME,true, 
         boost::bind(&UserEventLogTestCase::trigger_fired, this, _1, _2),
         boost::bind(&UserEventLogTestCase::trigger_inserted, this, _1, _2) ); 
} 

// callback indicating that trigger insertion is finished
void UserEventLogTestCase::trigger_inserted(const storage::Result &result,
                                           const storage::Trigger_id &tid) {
  
  if(result.code != storage::Result::SUCCESS) { 
    lg.err("inserting table trigger failed: %s \n", result.message.c_str());
    exit_test(false);  
    return; 
  } 
  ++trigger_count; 
  if(trigger_count == 1) {
    tid1 = tid; 
  } else if(trigger_count == 2) {
    tid2 = tid; 
  }else if(trigger_count == 3) { 
    tid3 = tid; 
    do_test(); 
  }
} 


void UserEventLogTestCase::do_test() { 
  if(debug) printf("running user_event_log c++ test\n"); 

  Log_entry_callback fn = boost::bind(&UserEventLogTestCase::log_entry_cb, 
                        this,_1,_2,_3,_4,_5,_6,_7); 
  string app_name = "c++ test"; 

  cur_msg = "log mesg 1";
  if(debug) printf("logging msg: %s \n", cur_msg.c_str());
  uel->log_simple(app_name,LogEntry::INFO, cur_msg); 
  test_sem->down(); 

  // expect nothing as far as names
  uel->get_log_entry(triggered_logid, fn);
  test_sem->down(); 

  // add mapping for MAC = 1 and MAC = 2
  b_store->store_binding_state(datapathid::from_host(1),1,1,1,
        "javelina",Name::HOST,false,0); 
  b_store->store_binding_state(datapathid::from_host(1),1,1,1,
      "dan",Name::USER,false,0); 
  b_store->store_binding_state(datapathid::from_host(2),2,2,2,
      "finland",Name::HOST,false, 0); 
  b_store->store_binding_state(datapathid::from_host(2),2,2,2,
      "teemu",Name::USER,false,0); 
  test_sem->down();
  test_sem->down();
  test_sem->down();
  test_sem->down();
 
  cur_msg = "MAC = 1 and MAC = 2 are BAD"; 
  if(debug) printf("logging msg: %s \n", cur_msg.c_str());
  LogEntry e1(app_name,LogEntry::INFO, cur_msg);
  e1.addMacKey(ethernetaddr(1),LogEntry::SRC); 
  e1.addMacKey(ethernetaddr(2),LogEntry::DST); 
  uel->log(e1); 
  test_sem->down();  // adds six name rows, one main row
  test_sem->down();
  test_sem->down();
  test_sem->down();
  test_sem->down();
  test_sem->down();
  test_sem->down();

  expected_src_names.clear();
  expected_src_names.push_back(Name("javelina",Name::HOST));
  expected_src_names.push_back(Name("dan",Name::USER));
  expected_src_names.push_back(Name("none;000000000001:1#none;000000000001#1",Name::LOCATION));
  expected_dst_names.clear();
  expected_dst_names.push_back(Name("finland",Name::HOST));
  expected_dst_names.push_back(Name("teemu",Name::USER));
  expected_dst_names.push_back(Name("none;000000000002:2#none;000000000002#2",Name::LOCATION));
  uel->get_log_entry(triggered_logid, fn);
  test_sem->down(); 
  
  cur_msg = "src IP = 1 is good, dest unknown";
  if(debug) printf("logging msg: %s \n", cur_msg.c_str());
  LogEntry e2(app_name,LogEntry::INFO, cur_msg);
  e2.addIPKey(1,LogEntry::SRC); 
  e2.addIPKey(0,LogEntry::DST); 
  uel->log(e2); 
  test_sem->down();
  test_sem->down();
  test_sem->down();
  test_sem->down();
  
  expected_src_names.clear();
  expected_src_names.push_back(Name("javelina",Name::HOST));
  expected_src_names.push_back(Name("dan",Name::USER));
  expected_src_names.push_back(Name("none;000000000001:1#none;000000000001#1",Name::LOCATION));
  expected_dst_names.clear();
  uel->get_log_entry(triggered_logid, fn);
  test_sem->down(); 
  
  cur_msg = "dst dpid = 2,port = 2 is good, src unknown";
  if(debug) printf("logging msg: %s \n", cur_msg.c_str());
  LogEntry e3(app_name,LogEntry::INFO, cur_msg);
  e3.addLocationKey(datapathid::from_host(0),0,LogEntry::SRC); 
  e3.addLocationKey(datapathid::from_host(2),2,LogEntry::DST); 
  uel->log(e3); 
  test_sem->down();
  test_sem->down();
  test_sem->down();
  test_sem->down();
  
  expected_src_names.clear();
  expected_dst_names.clear();
  expected_dst_names.push_back(Name("finland",Name::HOST));
  expected_dst_names.push_back(Name("teemu",Name::USER));
  expected_dst_names.push_back(Name("none;000000000002:2#none;000000000002#2",Name::LOCATION));
  uel->get_log_entry(triggered_logid, fn);
  test_sem->down(); 

  // done logging
  store->remove_trigger(tid1, 
        boost::bind(&UserEventLogTestCase::remove_trigger_callback,this,_1)); 
  store->remove_trigger(tid2, 
        boost::bind(&UserEventLogTestCase::remove_trigger_callback,this,_1)); 
  store->remove_trigger(tid3, 
        boost::bind(&UserEventLogTestCase::remove_trigger_callback,this,_1)); 

} 


void UserEventLogTestCase::log_entry_cb(int64_t logid, int64_t ts, 
                        const string &app, int level, const string &msg, 
                        const NameList &src_names, const NameList &dst_names){
  if(debug) { 
    printf("expected src names: \n"); 
    Bindings_Storage::print_names(expected_src_names); 
    printf("expected dst names: \n"); 
    Bindings_Storage::print_names(expected_dst_names); 
    printf("got src names: \n"); 
    Bindings_Storage::print_names(src_names); 
    printf("got dst names: \n"); 
    Bindings_Storage::print_names(dst_names); 
  } 

  BOOST_REQUIRE(src_names.size() == expected_src_names.size());  
  NameList::const_iterator it = src_names.begin(); 
  for( ; it != src_names.end(); it++) {  
    BOOST_REQUIRE(find(expected_src_names.begin(), expected_src_names.end(), 
          *it) != expected_src_names.end()); 
  }

  BOOST_REQUIRE(dst_names.size() == expected_dst_names.size());  
  it = dst_names.begin(); 
  for( ; it != dst_names.end(); it++) {  
    BOOST_REQUIRE(find(expected_dst_names.begin(), expected_dst_names.end(), 
          *it) != expected_dst_names.end()); 
  }


  test_sem->up(); 
} 

void UserEventLogTestCase::trigger_fired(const storage::Trigger_id & tid,
                                        const storage::Row& r) { 
  if(tid.ring == User_Event_Log::MAIN_TABLE_NAME) {
    try {
      // message inserted into table should be same one that we just logged 
      BOOST_REQUIRE(cur_msg == Storage_Util::get_col_as_type<string>(r,"msg"));
      triggered_logid = Storage_Util::get_col_as_type<int64_t>(r,"logid");
    }catch (exception &e) { 
      printf("%s \n", e.what());
      exit_test(false); 
      return;
    }
  } 
  test_sem->up(); 
} 


void UserEventLogTestCase::remove_trigger_callback(
                                    const storage::Result &result) { 
  if(result.code != storage::Result::SUCCESS) { 
    lg.err("removing table trigger failed: %s \n", result.message.c_str());
    exit_test(false); 
    return; 
  } 
  // exit test with success
  --trigger_count; 
  if(trigger_count == 0) 
    exit_test(true); 
} 

// use semephore to tell test runner that we are done
void UserEventLogTestCase::exit_test(bool success) { 
  if(!success) 
    BOOST_REQUIRE(false); // fail
  main_sem->up(); 
} 

void UserEventLogTestCase::clear_cb() { 
  test_sem->up(); // so we can wait for clear to finish 
} 


}  //unamed namespace 

BOOST_AUTO_COMPONENT_TEST_CASE(UserEventLogTest, UserEventLogTestCase);

File: nox/src/nox/netapps/tests/pyunittests/user_event_log_test.py
# Copyright 2008 (C) Nicira, Inc.
# 
# This file is part of NOX.
# 
# NOX is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# NOX is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with NOX.  If not, see <http://www.gnu.org/licenses/>.

from nox.lib.netinet.netinet import *
from nox.lib.core     import Component
from nox.netapps.storage import Storage
from nox.netapps.user_event_log.pyuser_event_log import pyuser_event_log,LogEntry
from nox.netapps.bindings_storage.pybindings_storage import pybindings_storage,Name
from nox.netapps.tests import unittest
from twisted.internet.defer import Deferred
import time

pyunit = __import__('unittest')

class UserEventLogTestCase(unittest.NoxTestCase):
    """Python test component to verify that python bindings
       work for the User_Event_Log component"""

    def configure(self,configuration):
      pass

    def install(self):
      self.counter = 0
      self.uel = self.resolve(pyuser_event_log)
      # disable auto-clear of UEL
      self.uel.set_max_num_entries(-1) 
      self.store = self.resolve(Storage)
      self.bs = self.resolve(pybindings_storage) 
      self.trigger_count = 0
      self.debug = False
#      self.debug = True

    def getInterface(self):
        return str(UserEventLogTestCase)

    def run_test(self):
      if(self.debug) : print "starting User_Event_Log python test"
      # clear the user-event-log tables from past tests
      self.uel.clear(self.clear_cb)
      
      # create a deferred object that we will use
      # to tell the testing framework when we are done
      self.deferred = Deferred()
      return self.deferred

    def clear_cb(self): 

      
      d = self.store.put_table_trigger("user_event_log", True, 
                                        self.uel_trigger_callback)
      d.addCallback(self.set_trigger_callback)  
      d = self.store.put_table_trigger("user_event_log_names", True, 
                                        self.uel_trigger_callback)
      d.addCallback(self.set_trigger_callback)  
      d = self.store.put_table_trigger("bindings_location", True, 
                                        self.other_trigger_callback)
      d.addCallback(self.set_trigger_callback)  
      d = self.store.put_table_trigger("bindings_name", True, 
                                        self.other_trigger_callback)
      d.addCallback(self.set_trigger_callback)  

    def get_message(self): 
      return "log msg # " + str(self.counter) 

    def exit_test(self, success) :
      if(not success):
        assert(False)
      self.deferred.callback("done") # tell test-suite that we are done

    def set_trigger_callback(self, res): 
      result , tid = res 

      if (result[0] != 0): 
        print "put trigger failed: " + str(result[1]) 
        self.exit_test(False) 
        return 
      self.trigger_count += 1
      if self.trigger_count == 1:
        self.tid1 = tid 
      if self.trigger_count == 2: 
        self.tid2 = tid
      if self.trigger_count == 3: 
        self.tid3 = tid
      if self.trigger_count == 4: 
        self.tid4 = tid
        # insert a host binding so we can test get_logids_for_name()
        # control flow continues within trigger_callback
        self.bs.store_binding_state(datapathid.from_host(0),0,
                        ethernetaddr(0),0,"javelina", Name.HOST, False,0) 
    
    def other_trigger_callback(self, trigger_id, row, reason):
      if trigger_id[1] == "bindings_name":
        # user/host binding has been added, let's add a location binding
        self.bs.add_name_for_location(datapathid.from_host(0),0,
                              "of6k-5", Name.LOCATION)
      elif trigger_id[1] == "bindings_location":
        # location binding has been added, let's start logging
        self.log_once()
   
    # triggers for 'user_event_log' and 'user_event_log_names' tables
    # used to trigger the call to get_log_entry() once all rows have been
    # written as indicated by the wait_count value
    def uel_trigger_callback(self, trigger_id, row, reason):
      if trigger_id[1] == "user_event_log":
        assert(row['msg'] == self.get_message())
     
      if(self.debug): print "changed row: " + str(row) 
      self.wait_count -= 1
      if(self.wait_count == 0): 
        if(self.debug): print "doing get_log_entry id = " + str(row['logid']) 
        self.uel.get_log_entry(int(row['logid']),self.get_entry_callback)
    
    def log_once(self):
      if(self.debug): print "starting log_once"
      self.counter = self.counter + 1
      msg = self.get_message()
      if self.counter == 1:
        self.wait_count = 1
        self.uel.log_simple("UEL Test Python",LogEntry.INFO,msg) 
      elif  self.counter == 2: 
        self.wait_count = 5
        dpid = datapathid.from_host(0)
        self.uel.log("UEL Test Python",LogEntry.INFO,msg,
                src_location=(dpid,0), dst_location=(dpid,0))
      elif self.counter == 3: 
        mac = ethernetaddr(0) 
        self.wait_count = 5
        self.uel.log("UEL Test Python",LogEntry.INFO,msg, 
                                    src_mac = mac, dst_mac = mac) 
      elif self.counter == 4: 
        self.wait_count = 5
        self.uel.log("UEL Test Python",LogEntry.INFO,msg, src_ip=0, dst_ip=0) 
      elif self.counter == 5:
        self.wait_count = 2
        self.uel.log("UEL Test Python",LogEntry.INFO,msg, sh="javelina" )
      elif self.counter == 6: 
        self.wait_count = 2
        dpid = datapathid.from_host(0)
        self.uel.log("UEL Test Python",LogEntry.INFO,msg, set_dst_loc=(dpid,0))
      else:
        # done inserting.  now test get_logids_for_name() 
        self.uel.get_logids_for_name("javelina",Name.HOST,
                                            self.get_logids_cb)
    def get_logids_cb(self, logids):
        assert(len(logids) == 4) 

        # all done with tests.  remove triggers
        d = self.store.remove_trigger(self.tid1)
        d.addCallback(self.remove_trigger_callback)
        d = self.store.remove_trigger(self.tid2)
        d.addCallback(self.remove_trigger_callback)
        d = self.store.remove_trigger(self.tid3)
        d.addCallback(self.remove_trigger_callback)
        d = self.store.remove_trigger(self.tid4)
        d.addCallback(self.remove_trigger_callback)
        
    def get_entry_callback(self,logid,ts,app,level,msg,src_names,dst_names):
      if self.debug : 
        print "src_names:"
        for n in src_names:
          print n
        print "dst_names:"
        for n in dst_names:
          print n

      assert(msg == self.get_message())
      if(self.counter == 1):  
        assert(len(src_names) == 0)
        assert(len(dst_names) == 0)
      elif (self.counter == 5): # log with no key, just a sh=javelina 
        assert(len(src_names) == 1)
        assert(len(dst_names) == 0)
      elif (self.counter == 6): # log with no key, just a dloc=(0,0)
        assert(len(src_names) == 0)
        assert(len(dst_names) == 1)
      else: 
        assert(len(src_names) == 2)
        assert(len(dst_names) == 2)

      # let's do it again!
      self.log_once() 

    def remove_trigger_callback(self, result): 
      self.trigger_count -= 1
      if (result[0] != 0): 
        print "remove trigger failed: " + str(result[1]) 
        self.exit_test(False) 
        return 
      elif self.trigger_count == 0:
        self.exit_test(True) # success 

def suite(ctxt):
    suite = pyunit.TestSuite()
    suite.addTest(UserEventLogTestCase("run_test", ctxt))
    return suite
